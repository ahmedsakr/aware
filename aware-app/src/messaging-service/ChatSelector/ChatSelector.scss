$vertical_menu_background_color: #18202c;

#vertical-menu {
    background-color: $vertical_menu_background_color;
    display: flex;
    flex-direction: column;
    text-align: left;
    height: 100%;

    >h3 {
        color: white;
        padding-left: 10px;
    }
}