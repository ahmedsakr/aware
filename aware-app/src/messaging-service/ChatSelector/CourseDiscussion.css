#vertical-menu {
    background-color: #18202c;
    display: block;
    text-align: left;
    height: 100%;
}

#vertical-menu > h3 {
  color: white;
  padding-left: 10px;
}

.chat-navigation-item {
    display: block;
    margin: 0px;
}

.navbar-item-name, .navbar-item-avatar {
  display: inline-block;
}

.navbar-item-name {
  color:white;
}

.chat-navigation-item:hover {
    background-color: #232f3e;
    text-decoration: none;
}

.chat-navigation-item.active {
    background-color: #4CAF50;
    color: white;
}

#tag:hover {
  background-color: #232f3e;
}

hr {
  background-color: white;
  height: 3;
}
