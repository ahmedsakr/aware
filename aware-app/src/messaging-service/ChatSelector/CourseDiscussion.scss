/*
 * Course Discussion styling
 *
 * Defines layout fundamentals for Course Discussion.
 */

$chat_navigation_background_color: #232f3e;
$chat_navigation_active_color: #4CAF50;
$navbar_item_color: #babdc0;

.chat-navigation-item {
    display: block;
    margin: 0px;
}

.chat-navigation-item-selected {
    display: block;
    margin: 0px;
    background-color: $chat_navigation_background_color;
}

.navbar-item-name,
.navbar-item-avatar {
    display: inline-block;
}

.navbar-item-name {
    color: $navbar_item_color;

    >p {
        margin: 0px;
    }
}

.chat-navigation-item {
    &:hover {
        background-color: $chat_navigation_background_color;
        text-decoration: none;
    }

    &.active {
        background-color: $chat_navigation_active_color;
        color: white;
    }
}

#tag:hover {
    background-color: $chat_navigation_background_color;
}

.chat-selector-line-break {
    background-color: white;
    height: 3;
    width: 100%;
}