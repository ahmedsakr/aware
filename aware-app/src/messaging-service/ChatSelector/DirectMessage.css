.direct-message {
    color: white;
    padding: 0;
    margin: 0;
}

.direct-message:hover {
    background-color: #232f3e;
    text-decoration: none;
}

.direct-message-avatar, .direct-message-content {
    display: inline-block;
}

.direct-message-content > p {
  padding: 0;
  margin: 0;
}

.direct-message-content {
    text-align: left;
    display: inline-block;
}
.direct-message-preview {
    border-radius: 5px;
    font-style: italic;
}

@media screen and (max-width: 576px) {
    .direct-message-content > p {
        display: none;
    }

    #vertical-menu > h3 {
        display: none;
    }

    .navbar-item-name {
        display: none;
    }
}

@media screen and (min-width: 576px) and (max-width: 768px) {
    .direct-message-content > p {
        font-size: 0.8em;
    }
}

@media screen and (min-width: 768px) and (max-width: 992px) {
    .direct-message-content > p {
        font-size: 2em;
    }
  }