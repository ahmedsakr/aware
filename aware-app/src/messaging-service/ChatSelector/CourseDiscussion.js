import React, {Component} from 'react';
import './CourseDiscussion.css'
class CourseDiscussion extends Component {
  render() {
    const {src, name} = this.props;

    return (
      <a href="/" class="chat-navigation-item">
        <div class="navbar-item-avatar col-2">
          <img src={process.env.PUBLIC_URL + src} alt={name}/>
        </div>
        <div class="navbar-item-name col-9">
          <p>{name.toUpperCase()}</p>
        </div>
      </a>
    );
  }
}

export default CourseDiscussion;
