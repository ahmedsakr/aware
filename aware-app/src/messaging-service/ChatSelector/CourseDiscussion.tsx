import React from 'react';
import './CourseDiscussion.scss'

type CourseDiscussionProps = {
    updateRoom: (course: CourseDiscussion) => void,
    room: string,
    src: string,
    name: string,
};

type CourseDiscussionState = {
    selected: boolean
};

export default class CourseDiscussion extends React.Component<CourseDiscussionProps, CourseDiscussionState> {
    constructor(props: CourseDiscussionProps) {
        super(props)

        this.state = {
            selected: false
        }
    }

    shouldComponentUpdate(nextProps: CourseDiscussionProps, nextState: CourseDiscussionState): boolean {
        return nextState.selected !== this.state.selected;
    }

    render(): JSX.Element {
        const { src, name, room, updateRoom } = this.props;

        return (
            <div onClick={() => { updateRoom(this); }} className={this.state.selected ? "chat-navigation-item-selected" : "chat-navigation-item"}>
                <div className="navbar-item-avatar col-2">
                    <img src={process.env.PUBLIC_URL + src} alt={name} />
                </div>
                <div className="navbar-item-name col-9">
                    <p>{name}</p>
                </div>
            </div>
        );
    }
}
