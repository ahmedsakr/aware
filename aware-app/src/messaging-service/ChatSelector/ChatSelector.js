import React, {Component} from 'react';
import CourseDiscussion from './CourseDiscussion'
import './CourseDiscussion.css';
import DirectMessage from './DirectMessage/DirectMessage'

class ChatSelector extends Component {
  constructor() {
    super()
    
    this.state = {
      rooms: [],
      selectedRoom: null
    };
  }

  componentWillMount() {
    if (this.props.socket) {

      // Retrieve all rooms that the user is subscribed to.
      this.props.socket.emit('get-rooms', this.props.username);

      // Listen for any updates in subscribed rooms for this user.
      this.props.socket.on('user-rooms', (rooms) => {
        this.setState({
          rooms: rooms
        });
      });
    }
  }

  updateSelectedRoom(room) {
    this.setState({
      selectedRoom: room
    });
  }

  shouldComponentUpdate(nextProps, nextState) {
    return  (this.state.rooms !== nextState.rooms) || (this.state.selectedRoom !== nextState.selectedRoom);
  }

  componentDidUpdate(prevProps, prevState, snapshot) {
    
    if (this.state.selectedRoom) {
      this.props.selectRoom(this.state.selectedRoom.props.room, this.state.selectedRoom.props.name);
      this.state.selectedRoom.setState({ selected: true });
    }
 
    if (prevState.selectedRoom) {
      prevState.selectedRoom.setState({ selected: false });
    }
  }

  render() {

    return (
      <div id="vertical-menu">
        <h3>Course Discussion</h3>

        {
          this.state.rooms.map(room => {
            return (
              <CourseDiscussion
                updateRoom={this.updateSelectedRoom.bind(this)}
                room={room.group_id}
                src={"/messenger-icons/" + room.icon}
                name={room.name} />
            )
          })
        }

        <hr></hr>

        <h3>Direct Messages</h3>
        <DirectMessage src="/messenger-icons/account.png" name="Josh Campitelli" />
        <DirectMessage src="/messenger-icons/account.png" name="Ahmed Sakr" />
        <DirectMessage src="/messenger-icons/account.png" name="Arsalan Sadiq" />

      </div>
    );
  }
}

export default ChatSelector;
