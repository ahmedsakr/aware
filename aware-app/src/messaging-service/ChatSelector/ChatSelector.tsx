import React from 'react';
import CourseDiscussion from './CourseDiscussion'
import Courses from './Courses/Courses'
import './ChatSelector.scss'
import './CourseDiscussion.scss'
import DirectMessage from './DirectMessage/DirectMessage'
import DirectMessages from './DirectMessages/DirectMessages'

export type Room = React.Component<CourseDiscussion['props'] | DirectMessage['props']> | null
type ChatSelectorProps = {
    requestRoom: (roomName: string, username: string) => void,
    socket: SocketIOClient.Socket,
    username: string
};

type ChatSelectorState = {
    selectedRoom: Room
};

export default class ChatSelector extends React.Component<ChatSelectorProps, ChatSelectorState> {

    constructor(props: ChatSelectorProps) {
        super(props)

        this.state = {
            selectedRoom: null
        };
    }

    /**
     * Requests a room change from the server and updates the highlighted
     * room to reflect the new room.
     *
     * @param room A new course or direct message chosen by the user
     */
    selectChat(room: Room): void {
        if (this.state.selectedRoom !== room) {

            // De-select the joined room in preparation for joining the new room.
            if (this.state.selectedRoom) {
                this.state.selectedRoom.setState({ selected: false });
            }

            this.setState({
                selectedRoom: room
            }, () => {
                if (this.state.selectedRoom) {
                    this.state.selectedRoom.setState({ selected: true });

                    const { room, name } = this.state.selectedRoom.props;
                    this.props.requestRoom(room, name);                    
                }
            });
        }
    }

    /**
     * The ChatSelector should never have to update; instead, it just maintains the
     * current selected room state without a need for updating.
     */
    shouldComponentUpdate(nextProps: ChatSelectorProps, nextState: ChatSelectorState): boolean {
        return false;
    }

    render(): JSX.Element {

        return (
            <div id="vertical-menu">
                <h3>Course Discussion</h3>

                <Courses
                    socket={this.props.socket}
                    username={this.props.username}
                    selectCourse={this.selectChat.bind(this)} />

                <hr className="chat-selector-line-break" />

                <h3>Direct Messages</h3>

                <DirectMessages 
                    selectDirectMessage={this.selectChat.bind(this)}/>
            </div>
        );
    }
}
