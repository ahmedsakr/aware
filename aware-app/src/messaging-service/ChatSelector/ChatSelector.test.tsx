import React from 'react';
import { shallow } from 'enzyme';

import ChatSelector from './ChatSelector';
import io from 'socket.io-client'

jest.mock('socket.io-client');

describe('ChatSelector', () => {
    it('should render correctly', () => {

        // Stub out the function for now, but we probably need to
        // implement better tests in the future.
        let selectRoom = (roomName: string, username: string) => { };

        let channel = io();

        const component = shallow(
            <ChatSelector
                selectRoom={selectRoom}
                socket={channel}
                username="myusername" />
        );

        expect(component).toMatchSnapshot();
    });
});
