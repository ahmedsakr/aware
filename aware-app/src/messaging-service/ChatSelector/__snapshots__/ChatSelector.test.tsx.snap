version https://git-lfs.github.com/spec/v1
oid sha256:a078cd59ae344b00d1881619c0ec27c261f78ea2ab6e8ae14e16751c34bf638b
size 388
